### ISSUE TITLE
Implement support to Sanitize on v3

### ISSUE THREAD

**Issue by zimmerle** [SpiderLabs/ModSecurity#1898]
As listed at #715 the Sanitize actions are not yet working on v3.

https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#sanitiseArg

**Comment by zimmerle** [comment]
Typo fixed at: c2bc6952651707f773280b7b8188db814aa11718

**Comment by Veszner** [comment]
Hi, is there an update on this feature, when can we expect it? Is it currently planned for v3.1.1?

**Comment by martinhsv** [comment]
Hello @Veszner ,

The current plan is to implement some sanitization support in v3 a little later this year -- not in the currently-active v3.0.7, but in the v3 work immediately thereafter.

**Comment by AnnoyingTechnology** [comment]
Any news on this issue ?

**Comment by todd-richmond** [comment]
this keeps slipping, but is a blocker for anyone who requires password or other masking in logs. Can't upgrade from v2 w/o it

**Comment by martinhsv** [comment]
Hello @todd-richmond ,

Regarding "Can't upgrade from v2 w/o it" ...

The lack of sanitization capability is considered a gap in current v3 functionality -- and one that is desirable to fill.

But there is no particular reason to view migrating from ModSecurity v2 to ModSecurity v3 as important or highly desirable.

ModSecurity v2 continues to be the maintained version of the software that is suitable for use with Apache HTTP Server (and IIS), while ModSecurity v3 is the maintained version that is suitable for use with nginx.

**Comment by fardarter** [comment]
@martinhsv Do you not recommend people use 3.x?

**Comment by AnnoyingTechnology** [comment]
Any news on that matter ? 
Password (non)removal from logs is a serious security concern, especially for a security product.

**Comment by zadros4psa** [comment]
is there any plan for this to be solved?

**Comment by cello86** [comment]
Any news on this issue ??

**Comment by theseion** [comment]
Yes, we want to implement those operators, but we are short on resources, so we can't really say when. PRs are always welcome thouhgh.

**Comment by airween** [comment]
Hi @cello86,

> Any news on this issue ??

Not yet. Probably you saw the latest security issue at mod_security2, where the cause was the sanitize mechanism. We have to review that code (in v2) and figure out something new. After that we can add that to v3 too.

As @theseion wrote, we are very low on resources, but we can try to do as we can.

**Comment by ssigwart** [comment]
@theseion and @airween, my hack for this that's been working for years is to edit this code before compiling:
https://github.com/owasp-modsecurity/ModSecurity/blob/fc905d9ea49f3264eaf975b0d87985af741e7d15/src/operators/operator.cc#L107-L132

I change `utils::string::limitTo(100` to `utils::string::limitTo(4`.  That way, I should never see a full credit card PAN or password in the logs. It can be a little bit of a pain with debugging, but avoids the issues mentioned in this thread. I wonder if a setting can be added to enable something like that. If you like the idea, I can try to implement it. I could use a hint on how to start going about adding a true setting for it though.

**Comment by airween** [comment]
Hi @ssigwart,

> That way, I should never see a full credit card PAN or password in the logs. It can be a little bit of a pain with debugging, but avoids the issues mentioned in this thread. I wonder if a setting can be added to enable something like that. If you like the idea, I can try to implement it. I could use a hint on how to start going about adding a true setting for it though.

I'm not sure truncate an important information in the log in all cases is a good idea.

The sanitize mechanism works differently in the current v2 implementation. I try to explain very broadly it if you're interesting to add this feature to v3 (which would be a great help, indeed - and thanks for offering).

When a request is received by the engine, it extracts the argument from the request body. Consider a JSON request:
```JSON
{"key1":"val1","key2":"val2"}
```
When this extraction is happening, the engine stores the keys and their values' original position in the request body. [Here](https://github.com/owasp-modsecurity/ModSecurity/blob/v2/master/apache2/msc_json.c#L52-L56) you can see that part, the `origin_offset` variable holds the original position of the value, for eg. In the example above the offset value of `val1` will be `9`.

If a rule contains a `sanitize*` action, then it [adds](https://github.com/owasp-modsecurity/ModSecurity/blob/v2/master/apache2/re_actions.c#L1459) the name (the key) of the necessary argument to a table which we want to sanitize.

In the logging [phase](https://github.com/owasp-modsecurity/ModSecurity/blob/v2/master/apache2/msc_logging.c#L895), if the key exists in the given table, it looks for the already stored position in the extract method and replaces the value with `*` characters equal to the length of the already stored variable. (Indeed there are many more occurrences where the code checks that table.)

This is the same way in case of XML and www-urlencoded body, but again: this is a very broadly explanation about how it works.

As you can see, this is a bit complicated and not very elegant (but that's just my opinion). We should find a more elegant way to implement it, but I would be happy and supportive if someone could at least implement this solution.

One more note to your solution: you truncate the argument value (and/or the key too?) in function `resolveMatchMessages()`, which is used in error.log and audit.log too, but it's used in audit.log's `K` part. `sanitize*` action replaces the necessary values in audit.log's `C` section. With your solution the card numbers, passwords and other things will be still visible.

**Comment by ssigwart** [comment]
Thanks for the quick reply, @airween.

I tried looking at the v2 code and seeing how I might be able to add it to v3. I think the starting point might be in `RulesSet::evaluate` and it's call to `RuleWithOperator::evaluate`, but I haven't done C++ code in quite some time, so I probably won't be able to make much process on this. I haven't even figured out how to compile and run the tests.

> One more note to your solution: you truncate the argument value (and/or the key too?) in function `resolveMatchMessages()`, which is used in error.log and audit.log too, but it's used in audit.log's `K` part. `sanitize*` action replaces the necessary values in audit.log's `C` section. With your solution the card numbers, passwords and other things will be still visible.

I just do the value. The key is typically something that I define, so I'm not worried about that.

I set `SecAuditLogParts AIJDFHZ`, so the `C` section isn't in there.
